# Nasonia
Lack of parent-of-origin effects in Nasonia jewel wasp - a replication and reproduction study

## Background
Parent and species of origin allele specific and differential expression analysis workflow. A replication and reproduction study using RNA and DNA from Nasonia jewel wasp a haplodiploidy species. In diploid cells, the paternal and maternal alleles are on average equally expressed. There are exceptions from this is which a small number of genes express exclusively the maternal or paternal allele copy, known as genomic imprinting. Clark et al. 2016 found no parent-of-origin effects in the hybrids of closely related Nasonia vitripennis and N. giuralti jewel wasp, suggesting a lack of epigenetic reprogramming during embryogenesis in these species. Here, we have reproduced and replicated these findings using the previously published RNA & DNA sequence data from 11 samples as well as a newly generated RNA data from 12 samples of the same species hybrids. Our results from both datasets demonstrated a species-of-origin effect.

## Preprint
Please see our preprint for more information:

*Lack of parent-of-origin effects in Nasonia jewel wasp - a replication and reproduction study* 2019. Olney KC; Natri H; Underwood A; Gibson J; Gadau J; Wilson MA.  

If you use Nasonia in a classroom or discuss/correct this work in a manuscript, please cite this preprint.

## Contents:
1. Process data - Download and process files and identify sites that are fixed and different between the inbreed lines of N. Vit and N. Gir 

2. Differential expression - Differential expression for all pairwise comparisons 

3. Allele specific expression - Identify genes with species or parent-of-origin effects

### 1. Process data

### 2. Differential exprssion

### 3. Allele specific expression 

### publicly available tools used in this analysis
Tool | usage | citation
--- | --- |  ---
Trimmomatic | Trim RNA-sequences for quality | Bolger AM, Lohse M, Usadel B. Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics. 2014;30: 2114–2120.
HISAT | RNAseq read aligner | Kim D, Langmead B, Salzberg SL. HISAT: a fast spliced aligner with low memory requirements. Nat Methods. 2015;12: 357–360.
bamtools | analyzing and processing BAM files | Barnett DW, Garrison EK, Quinlan AR, Strömberg MP, Marth GT. BamTools: a C++ API and toolkit for analyzing and managing BAM files. Bioinformatics. 2011;27: 1691–1692.
FeatureCounts | obtain raw transcriptome counts| Liao Y, Smyth GK, Shi W. featureCounts: an efficient general purpose program for assigning sequence reads to genomic features. Bioinformatics. 2014;30: 923–930.
Limma/voom | differenital expression analysis | Law CW, Chen Y, Shi W, Smyth GK. voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biol. 2014;15: R29.


## Group Members
Name | email | github ID
--- | --- |  ---
Kimberly Olney | olneykimberly@gmail.com | @olneykimberly
Heini Natri | heini.natri@gmail.com |@heinin
Melissa A. Wilson | melissa.wilsonsayres@asu.edu | @mwilsonsayres
